File: src/main/kotlin/repositories/utils/StudentFilterUtils.kt
package repositories.utils

import org.jetbrains.exposed.v1.core.or
import org.jetbrains.exposed.v1.jdbc.Query
import org.jetbrains.exposed.v1.jdbc.andWhere
import schema.model.OrderStudentsByClause
import schema.model.OrderStudentsByColumn
import schema.model.PaginationArgs
import schema.model.SortDirection
import schema.model.StudentFilterArgs
import tables.Students
import kotlin.collections.forEach

/**
 * Applies filtering to a student query based on the provided filter arguments.
 *
 * @param filterArgs The filter arguments to apply.
 * @return The query with the filters applied.
 */
fun Query.applyStudentFilters(filterArgs: StudentFilterArgs?): Query = also {
    filterArgs?.let { args ->
        val ftsLang = "simple" // Use "simple" for language-agnostic, or "english", "arabic" etc.

        // Name filters
        args.name?.let {
            if (it.isNotBlank()) {
                andWhere { TsVector(Students.name, ftsLang) matches TsQuery(it, ftsLang) }
            }
        }
        args.nameNotContains?.let { andWhere { Students.name notLike "%$it%" } }
        args.nameEquals?.let { andWhere { Students.name eq it } }
        args.nameNotEquals?.let { andWhere { Students.name neq it } }
        args.nameStartsWith?.let { andWhere { Students.name like "$it%" } }
        args.nameEndsWith?.let { andWhere { Students.name like "%$it" } }
        args.nameIsEmpty?.let { if (it) andWhere { Students.name eq "" } } // Name is not nullable
        args.nameIsNotEmpty?.let { if (it) andWhere { Students.name neq "" } }

        // Email filters
        args.email?.let {
            if (it.isNotBlank()) {
                andWhere { TsVector(Students.email, ftsLang) matches TsQuery(it, ftsLang) }
            }
        }
        args.emailNotContains?.let { andWhere { Students.email notLike "%$it%" } }
        args.emailEquals?.let { andWhere { Students.email eq it.value } }
        args.emailNotEquals?.let { andWhere { Students.email neq it } }
        args.emailStartsWith?.let { andWhere { Students.email like "$it%" } }
        args.emailEndsWith?.let { andWhere { Students.email like "%$it" } }
        args.emailIsEmpty?.let {
            if (it) andWhere { Students.email.isNull() or (Students.email eq "") }
        }
        args.emailIsNotEmpty?.let { if (it) andWhere { Students.email.isNotNull() } }

        // Phone number filter
        args.phoneNumber?.let { andWhere { Students.phoneNumber eq it.number } }

        // Gender filter
        args.gender?.let { andWhere { Students.gender eq it } }

        // Nationality filter
        args.nationality?.let { andWhere { Students.nationality eq it } }

        // Created at filters
        args.createdAt?.let { andWhere { Students.createdAt eq it } }
        args.createdAtNot?.let { andWhere { Students.createdAt neq it } }
        args.createdAtFrom?.let { andWhere { Students.createdAt greaterEq it } }
        args.createdAtTo?.let { andWhere { Students.createdAt lessEq it } }
        args.createdAtAfter?.let { andWhere { Students.createdAt greater it } }
        args.createdAtBefore?.let { andWhere { Students.createdAt less it } }
        args.createdAtOnOrAfter?.let { andWhere { Students.createdAt greaterEq it } }
        args.createdAtOnOrBefore?.let { andWhere { Students.createdAt lessEq it } }
        args.createdAtIsEmpty?.let { if (it) andWhere { Students.createdAt.isNull() } }
        args.createdAtIsNotEmpty?.let { if (it) andWhere { Students.createdAt.isNotNull() } }

        // Birthdate filters (dateOfBirth is LocalDate, birthDate fields are LocalDateTime)
        args.birthDate?.let { andWhere { Students.dateOfBirth eq it.date } }
        args.birthDateNot?.let { andWhere { Students.dateOfBirth neq it.date } }
        args.birthDateFrom?.let { andWhere { Students.dateOfBirth greaterEq it.date } }
        args.birthDateTo?.let { andWhere { Students.dateOfBirth lessEq it.date } }
        args.birthDateAfter?.let { andWhere { Students.dateOfBirth greater it.date } }
        args.birthDateBefore?.let { andWhere { Students.dateOfBirth less it.date } }
        args.birthDateOnOrAfter?.let { andWhere { Students.dateOfBirth greaterEq it.date } }
        args.birthDateOnOrBefore?.let { andWhere { Students.dateOfBirth lessEq it.date } }
        args.birthDateIsEmpty?.let { if (it) andWhere { Students.dateOfBirth.isNull() } }
        args.birthDateIsNotEmpty?.let { if (it) andWhere { Students.dateOfBirth.isNotNull() } }
    }
}

/**
 * Applies ordering to a student query based on the provided order by clauses.
 *
 * @param orderBy The list of order by clauses.
 * @return The query with ordering applied.
 */
fun Query.applyStudentOrdering(orderBy: List<OrderStudentsByClause>?): Query = also {
    orderBy?.forEach { clause ->
        val sortDirection = when (clause.order) {
            SortDirection.ASC -> org.jetbrains.exposed.v1.core.SortOrder.ASC
            SortDirection.DESC -> org.jetbrains.exposed.v1.core.SortOrder.DESC
        }

        when (clause.column) {
            OrderStudentsByColumn.ID -> orderBy(Students.id, sortDirection)
            OrderStudentsByColumn.NAME -> orderBy(Students.name, sortDirection)
            OrderStudentsByColumn.EMAIL -> orderBy(Students.email, sortDirection)
            OrderStudentsByColumn.DATE_OF_BIRTH -> orderBy(Students.dateOfBirth, sortDirection)

            OrderStudentsByColumn.GENDER -> orderBy(Students.gender, sortDirection)
            OrderStudentsByColumn.CREATED_AT -> orderBy(Students.createdAt, sortDirection)
            OrderStudentsByColumn.UPDATED_AT -> orderBy(Students.updatedAt, sortDirection)
        }
    }
}

fun Query.paginate(args: PaginationArgs): Query = also {
    limit(args.perPage).offset(args.offset.toLong())
}

File: server/utils/student.filterUtils.ts
import * as Types from "@/server/types";
import { OrderSortDirection } from "@/lib/enum";
import { students } from "@/server/db/schema";
import {
    eq,
    notLike,
    like,
    not,
    or,
    isNull,
    gt,
    gte,
    lt,
    lte,
    asc,
    desc,
} from "drizzle-orm";
import { PgSelect } from "drizzle-orm/pg-core";
import { fullTextSearch } from "@/server/db/query.extentions";

export namespace StudentFilterUtils {
    export function applyFilters<T extends PgSelect>(
        qb: T,
        args?: Types.StudentFilterArgs | null,
    ) {
        if (!args) return qb;
        const ftsLang = "simple"; // Use "simple" for language-agnostic, or "english", "arabic" etc.

        // Name filters
        if (args.name) {
            const ftsCondition = fullTextSearch(
                students.name,
                args.name,
                ftsLang,
            );
            if (ftsCondition) {
                qb = qb.where(ftsCondition);
            }
        }
        if (args.nameNotContains) {
            qb = qb.where(notLike(students.name, `%${args.nameNotContains}%`));
        }
        if (args.nameEquals) {
            qb = qb.where(eq(students.name, args.nameEquals));
        }
        if (args.nameNotEquals) {
            qb = qb.where(not(eq(students.name, args.nameNotEquals)));
        }
        if (args.nameStartsWith) {
            qb = qb.where(like(students.name, `${args.nameStartsWith}%`));
        }
        if (args.nameEndsWith) {
            qb = qb.where(like(students.name, `%${args.nameEndsWith}`));
        }
        if (args.nameIsEmpty) {
            qb = qb.where(eq(students.name, ""));
        }
        if (args.nameIsNotEmpty) {
            qb = qb.where(not(eq(students.name, "")));
        }

        // Email filters (contains)
        if (args.email) {
            qb = qb.where(like(students.email, `%${args.email}%`));
        }
        if (args.emailNotContains) {
            qb = qb.where(
                notLike(students.email, `%${args.emailNotContains}%`),
            );
        }
        if (args.emailEquals?.value) {
            qb = qb.where(eq(students.email, args.emailEquals.value));
        }
        if (args.emailNotEquals) {
            qb = qb.where(not(eq(students.email, args.emailNotEquals)));
        }
        if (args.emailStartsWith) {
            qb = qb.where(like(students.email, `${args.emailStartsWith}%`));
        }
        if (args.emailEndsWith) {
            qb = qb.where(like(students.email, `%${args.emailEndsWith}`));
        }
        if (args.emailIsEmpty) {
            qb = qb.where(or(eq(students.email, ""), isNull(students.email)));
        }
        if (args.emailIsNotEmpty) {
            qb = qb.where(not(isNull(students.email)));
        }

        // Phone number filter (contains)
        if (args.phoneNumber?.number) {
            qb = qb.where(
                like(students.phoneNumber, `%${args.phoneNumber.number}%`),
            );
        }

        // Gender filter
        if (args.gender) {
            qb = qb.where(eq(students.gender, args.gender));
        }

        // Nationality filter
        if (args.nationality) {
            qb = qb.where(eq(students.nationality, args.nationality));
        }

        // Created at filters
        if (args.createdAt) {
            qb = qb.where(eq(students.createdAt, args.createdAt));
        }
        if (args.createdAtNot) {
            qb = qb.where(not(eq(students.createdAt, args.createdAtNot)));
        }
        if (args.createdAtFrom) {
            qb = qb.where(gte(students.createdAt, args.createdAtFrom));
        }
        if (args.createdAtTo) {
            qb = qb.where(lte(students.createdAt, args.createdAtTo));
        }
        if (args.createdAtAfter) {
            qb = qb.where(gt(students.createdAt, args.createdAtAfter));
        }
        if (args.createdAtBefore) {
            qb = qb.where(lt(students.createdAt, args.createdAtBefore));
        }
        if (args.createdAtOnOrAfter) {
            qb = qb.where(gte(students.createdAt, args.createdAtOnOrAfter));
        }
        if (args.createdAtOnOrBefore) {
            qb = qb.where(lte(students.createdAt, args.createdAtOnOrBefore));
        }
        if (args.createdAtIsEmpty) {
            qb = qb.where(isNull(students.createdAt));
        }
        if (args.createdAtIsNotEmpty) {
            qb = qb.where(not(isNull(students.createdAt)));
        }

        // Birthdate filters (dateOfBirth is Date)
        if (args.birthDate) {
            qb = qb.where(eq(students.dateOfBirth, args.birthDate));
        }
        if (args.birthDateNot) {
            qb = qb.where(not(eq(students.dateOfBirth, args.birthDateNot)));
        }
        if (args.birthDateFrom) {
            qb = qb.where(gte(students.dateOfBirth, args.birthDateFrom));
        }
        if (args.birthDateTo) {
            qb = qb.where(lte(students.dateOfBirth, args.birthDateTo));
        }
        if (args.birthDateAfter) {
            qb = qb.where(gt(students.dateOfBirth, args.birthDateAfter));
        }
        if (args.birthDateBefore) {
            qb = qb.where(lt(students.dateOfBirth, args.birthDateBefore));
        }
        if (args.birthDateOnOrAfter) {
            qb = qb.where(gte(students.dateOfBirth, args.birthDateOnOrAfter));
        }
        if (args.birthDateOnOrBefore) {
            qb = qb.where(lte(students.dateOfBirth, args.birthDateOnOrBefore));
        }
        if (args.birthDateIsEmpty) {
            qb = qb.where(isNull(students.dateOfBirth));
        }
        if (args.birthDateIsNotEmpty) {
            qb = qb.where(not(isNull(students.dateOfBirth)));
        }

        return qb;
    }

    export function applyOrdering<T extends PgSelect>(
        qb: T,
        orderBy?: Types.StudentsOrderByClause[] | null,
    ) {
        if (!orderBy || orderBy.length === 0) return qb;

        orderBy.forEach((clause) => {
            let column;
            switch (clause.column) {
                case Types.StudentsOrderByColumn.ID:
                    column = students.id;
                    break;
                case Types.StudentsOrderByColumn.NAME:
                    column = students.name;
                    break;
                case Types.StudentsOrderByColumn.EMAIL:
                    column = students.email;
                    break;
                case Types.StudentsOrderByColumn.DATE_OF_BIRTH:
                    column = students.dateOfBirth;
                    break;
                case Types.StudentsOrderByColumn.GENDER:
                    column = students.gender;
                    break;
                case Types.StudentsOrderByColumn.CREATED_AT:
                    column = students.createdAt;
                    break;
                case Types.StudentsOrderByColumn.UPDATED_AT:
                    column = students.updatedAt;
                    break;
                default:
                    return;
            }
            qb = qb.orderBy(
                clause.order === OrderSortDirection.ASC
                    ? asc(column)
                    : desc(column),
            );
        });
        return qb;
    }
}
